### PR TITLE
Implement ads library support for Brave Search Ads

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/20852

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See spec

### CODE REVIEW

**Comment by tmancey** [comment]
@moritzhaller @aseren search result ads were copied from another ad type. I will raise a ticket to refactor duplicate code for different ad types

**Review by kylehickinson — APPROVED** [review]
iOS lgtm

**Review by moritzhaller — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad.cc [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            HasFiredAdViewedEvent(ad, ad_events)) {
+          BLOG(1,
+               "Search result ad: Not allowed as already viewed uuid " << uuid);
+          NotifySearchResultAdEventFailed(uuid, creative_instance_id,
+                                          event_type);
+          return;
+        }
+
+        if (event_type == mojom::SearchResultAdEventType::kViewed) {
+          // TODO(tmancey): We need to fire an ad served event until promoted
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad.cc]: Fixed, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/ad_events/search_result_ads/search_result_ad_event_served.cc [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  LogAdEvent(ad, ConfirmationType::kServed, [](const bool success) {
+    if (!success) {
+      BLOG(1, "Failed to log search result ad served event");
+      return;
+    }
+
+    BLOG(6, "Successfully logged search result ad served event");
+  });
+}
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_events/search_result_ads/search_result_ad_event_served.cc:33]: do we need to add adding to history here? `history::AddSearchResultAd`

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ad_events/search_result_ads/search_result_ad_event_served.cc:33]: We only add to the history when the ad is viewed. Search ads will not be served by the library for the MVP so once a view event is fired a served event is automatically fired. Thanks

File: vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc [hunk 3]
```diff
@@ -88,5 +89,13 @@ void AddInlineContentAd(const InlineContentAdInfo& ad,
   Client::Get()->AppendAdHistory(ad_history);
 }
 
+void AddSearchResultAd(const SearchResultAdInfo& ad,
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc:92]: why are we adding new functions with similar functionality? should we just use one function, i.e. `AddAdHistory`

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc:92]: I have a list of tech debt issues to remove duplicate code, I will triage this soon

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/creative_search_result_ad_info.cc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+CreativeSearchResultAdInfo::CreativeSearchResultAdInfo(
+    const CreativeAdInfo& creative_ad)
+    : CreativeAdInfo(creative_ad) {}
+
+CreativeSearchResultAdInfo::~CreativeSearchResultAdInfo() = default;
+
+bool CreativeSearchResultAdInfo::operator==(
+    const CreativeSearchResultAdInfo& rhs) const {
+  return title == rhs.title && body == rhs.body;
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/bundle/creative_search_result_ad_info.cc:23]: should we compare fields of the parent class?

File: vendor/bat-native-ads/src/bat/ads/internal/catalog/catalog_creative_search_result_ad_info.cc [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+CatalogCreativeSearchResultAdInfo::CatalogCreativeSearchResultAdInfo(
+    const CatalogCreativeSearchResultAdInfo& info) = default;
+
+CatalogCreativeSearchResultAdInfo::~CatalogCreativeSearchResultAdInfo() =
+    default;
+
+bool CatalogCreativeSearchResultAdInfo::operator==(
+    const CatalogCreativeSearchResultAdInfo& rhs) const {
+  return payload == rhs.payload;
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/catalog/catalog_creative_search_result_ad_info.cc:21]: should we compare fields of the parent class

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_hour_permission_rule.cc [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  return true;
+}
+
+std::string SearchResultAdsPerHourPermissionRule::GetLastMessage() const {
+  return last_message_;
+}
+
+bool SearchResultAdsPerHourPermissionRule::DoesRespectCap(
+    const std::deque<base::Time>& history) {
+  const base::TimeDelta& time_constraint = base::Hours(1);
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_hour_permission_rule.cc:41]: nit: reference not needed

**Comment by aseren** [comment]
LGTM, except several nits

**Review by aseren — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_day_permission_rule.cc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  return true;
+}
+
+std::string SearchResultAdsPerDayPermissionRule::GetLastMessage() const {
+  return last_message_;
+}
+
+bool SearchResultAdsPerDayPermissionRule::DoesRespectCap(
+    const std::deque<base::Time>& history) {
+  const base::TimeDelta& time_constraint = base::Days(1);
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_day_permission_rule.cc]: Resolved with https://github.com/brave/brave-browser/issues/21881

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_hour_permission_rule.cc (hunk 6 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_hour_permission_rule.cc]: Resolved with https://github.com/brave/brave-browser/issues/21881

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_day_permission_rule.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ public:
+  SearchResultAdsPerDayPermissionRule();
+  ~SearchResultAdsPerDayPermissionRule() override;
+
+  bool ShouldAllow() override;
+
+  std::string GetLastMessage() const override;
+
+ private:
+  std::string last_message_;
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/permission_rules/search_result_ads_per_day_permission_rule.h:30]: Raised https://github.com/brave/brave-browser/issues/21929

File: vendor/bat-native-ads/src/bat/ads/internal/bundle/creative_search_result_ad_info.cc (hunk 4 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/bundle/creative_search_result_ad_info.cc:23]: Raised https://github.com/brave/brave-browser/issues/21928

File: vendor/bat-native-ads/src/bat/ads/internal/account/deposits/deposit_info.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <string>
+
+#include "base/time/time.h"
+
+namespace ads {
+
+struct DepositInfo final {
+  DepositInfo();
+  DepositInfo(const DepositInfo& info);
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/account/deposits/deposit_info.h:17]: if copy constructor is added then copy operator should be also defined

File: vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+  // Act
+  search_result_ad_->FireEvent(ad_mojom,
+                               mojom::SearchResultAdEventType::kViewed);
+
+  // Assert
+  EXPECT_TRUE(did_serve_ad_);
+  EXPECT_TRUE(did_view_ad_);
+  EXPECT_FALSE(did_click_ad_);
+  EXPECT_FALSE(did_fail_to_fire_event_);
+  const SearchResultAdInfo& expected_ad = BuildSearchResultAd(ad_mojom);
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc:164]: nit: extra reference

File: vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc:188]: nit: extra reference
and at some other places here

File: vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc [hunk 12]
```diff
@@ -88,5 +89,13 @@ void AddInlineContentAd(const InlineContentAdInfo& ad,
...
+                       const ConfirmationType& confirmation_type) {
+  const AdHistoryInfo& ad_history =
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc:94]: nit: extra reference

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 13]
```diff
@@ -567,6 +586,9 @@ void AdsImpl::set(privacy::TokenGeneratorInterface* token_generator) {
   promoted_content_ad_ = std::make_unique<PromotedContentAd>();
   promoted_content_ad_->AddObserver(this);
 
+  search_result_ad_ = std::make_unique<SearchResultAd>();
+  search_result_ad_->AddObserver(this);
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:590]: I didn't find where RemoveObserver is called

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:590]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc (hunk 12 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_history/ads_history.cc:94]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc (hunk 11 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc:188]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc (hunk 10 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/search_result_ads/search_result_ad_unittest.cc:164]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/account/deposits/deposit_info.h (hunk 9 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/account/deposits/deposit_info.h:17]: Removed as not required and raised https://github.com/brave/brave-browser/issues/22245 to fix elsewhere

**Pull request merged**